### PR TITLE
[TorchFT] Extract TorchFT checkpoint logic from core into experiments/ft/

### PR DESCRIPTION
**Pull request by mori360**
## Summary
  - Move all TorchFT fault-tolerance checkpoint logic out of core `CheckpointManager` into a new `FTCheckpointManager` subclass in                         
  `torchtitan/experiments/ft/checkpoint.py`                                                                                       
  - Extract `_should_purge()` as an overridable hook so the FT subclass can add `participating_rank` guards without duplicating the purge loop             
  - Remove FT-specific test fixtures from core unit tests

## Test plan
Test the main checkpoint by training 50 steps first, save, then load and train the next 50 steps.
```
$ NGPU=4 LOG_RANK=0 ./run_train.sh --checkpoint.enable --checkpoint.interval 50 --training.steps 50 --debug.seed 42 --debug.deterministic  --lr_scheduler.total_steps 100 
[rank0]:[titan] 2026-04-06 10:32:53,780 - root - INFO - step:  1  loss:  8.14122  grad_norm:  1.5181  memory:  0.91GiB(0.95%)  tps: 3,632  tflops: 0.26  mfu: 0.03%
[rank0]:[titan] 2026-04-06 10:32:54,517 - root - INFO - step: 10  loss:  3.58616  grad_norm:  1.3104  memory:  0.96GiB(1.01%)  tps: 193,943  tflops: 13.88  mfu: 1.40%
[rank0]:[titan] 2026-04-06 10:32:55,328 - root - INFO - step: 20  loss:  3.01840  grad_norm:  0.4317  memory:  0.96GiB(1.01%)  tps: 196,353  tflops: 14.06  mfu: 1.42%
[rank0]:[titan] 2026-04-06 10:32:56,103 - root - INFO - step: 30  loss:  2.86744  grad_norm:  0.4638  memory:  0.96GiB(1.01%)  tps: 194,395  tflops: 13.92  mfu: 1.41%
[rank0]:[titan] 2026-04-06 10:32:56,812 - root - INFO - step: 40  loss:  2.86680  grad_norm:  0.3045  memory:  0.96GiB(1.01%)  tps: 241,692  tflops: 17.30  mfu: 1.75%
[rank0]:[titan] 2026-04-06 10:32:57,549 - root - INFO - step: 50  loss:  2.80785  grad_norm:  0.2986  memory:  0.96GiB(1.01%)  tps: 172,155  tflops: 12.32  mfu: 1.25%


$ NGPU=4 LOG_RANK=0 ./run_train.sh --checkpoint.enable --checkpoint.interval 50 --training.steps 100 --debug.seed 42 --debug.deterministic 
[rank0]:[titan] 2026-04-06 10:33:25,626 - root - INFO - step: 60  loss:  2.82947  grad_norm:  0.2331  memory:  0.96GiB(1.01%)  tps: 234,383  tflops: 16.78  mfu: 1.70%
[rank0]:[titan] 2026-04-06 10:33:26,366 - root - INFO - step: 70  loss:  2.76000  grad_norm:  0.2966  memory:  0.96GiB(1.01%)  tps: 209,703  tflops: 15.01  mfu: 1.52%
[rank0]:[titan] 2026-04-06 10:33:27,123 - root - INFO - step: 80  loss:  2.77190  grad_norm:  0.3109  memory:  0.96GiB(1.01%)  tps: 216,492  tflops: 15.50  mfu: 1.57%
[rank0]:[titan] 2026-04-06 10:33:27,923 - root - INFO - step: 90  loss:  2.70831  grad_norm:  0.2125  memory:  0.96GiB(1.01%)  tps: 225,660  tflops: 16.15  mfu: 1.63%
[rank0]:[titan] 2026-04-06 10:33:28,667 - root - INFO - step: 100  loss:  2.75107  grad_norm:  0.1844  memory:  0.96GiB(1.01%)  tps: 193,858  tflops: 13.88  mfu: 1.40%

$ # remove checkpoint an rerun 100 steps
$ NGPU=4 LOG_RANK=0 ./run_train.sh --checkpoint.enable --checkpoint.interval 50 --training.steps 100 --debug.seed 42 --debug.deterministic
[rank0]:[titan] 2026-04-06 10:34:51,026 - root - INFO - step:  1  loss:  8.14122  grad_norm:  1.5181  memory:  0.91GiB(0.95%)  tps: 2,590  tflops: 0.19  mfu: 0.02%
[rank0]:[titan] 2026-04-06 10:34:51,741 - root - INFO - step: 10  loss:  3.58616  grad_norm:  1.3104  memory:  0.96GiB(1.01%)  tps: 221,129  tflops: 15.83  mfu: 1.60%
[rank0]:[titan] 2026-04-06 10:34:52,469 - root - INFO - step: 20  loss:  3.01840  grad_norm:  0.4317  memory:  0.96GiB(1.01%)  tps: 228,821  tflops: 16.38  mfu: 1.66%
[rank0]:[titan] 2026-04-06 10:34:53,210 - root - INFO - step: 30  loss:  2.86744  grad_norm:  0.4638  memory:  0.96GiB(1.01%)  tps: 202,643  tflops: 14.51  mfu: 1.47%
[rank0]:[titan] 2026-04-06 10:34:53,921 - root - INFO - step: 40  loss:  2.86680  grad_norm:  0.3045  memory:  0.96GiB(1.01%)  tps: 233,585  tflops: 16.72  mfu: 1.69%
[rank0]:[titan] 2026-04-06 10:34:54,698 - root - INFO - step: 50  loss:  2.80785  grad_norm:  0.2986  memory:  0.96GiB(1.01%)  tps: 177,377  tflops: 12.70  mfu: 1.28%
[rank0]:[titan] 2026-04-06 10:34:56,557 - root - INFO - step: 60  loss:  2.82947  grad_norm:  0.2331  memory:  0.96GiB(1.01%)  tps: 213,936  tflops: 15.31  mfu: 1.55%
[rank0]:[titan] 2026-04-06 10:34:57,282 - root - INFO - step: 70  loss:  2.76000  grad_norm:  0.2966  memory:  0.96GiB(1.01%)  tps: 232,720  tflops: 16.66  mfu: 1.68%
[rank0]:[titan] 2026-04-06 10:34:57,999 - root - INFO - step: 80  loss:  2.77190  grad_norm:  0.3109  memory:  0.96GiB(1.01%)  tps: 230,530  tflops: 16.50  mfu: 1.67%
[rank0]:[titan] 2026-04-06 10:34:58,741 - root - INFO - step: 90  loss:  2.70831  grad_norm:  0.2125  memory:  0.96GiB(1.01%)  tps: 233,548  tflops: 16.72  mfu: 1.69%
[rank0]:[titan] 2026-04-06 10:34:59,467 - root - INFO - step: 100  loss:  2.75107  grad_norm:  0.1844  memory:  0.96GiB(1.01%)  tps: 207,615  tflops: 14.86  mfu: 1.50%
```

### CODE REVIEW

**Comment by mori360** [comment]
@claude review this pr

**Comment by mori360** [comment]
> ### Nit: Pre-existing typo in folder name
> **`torchtitan/experiments/ft/checkpoint.py:193`**
> 
> ```python
> return os.path.join(self.folder, f"ft-replicat-{self.ft_replica_id}")
> ```
> 
> `ft-replicat-` is missing an 'e' — presumably should be `ft-replicate-` or `ft-replica-`. This is pre-existing from main so not blocking, but worth fixing since you're touching this code. Note that changing it would break loading existing FT checkpoints saved with the old name, so consider whether any users have such checkpoints.
> 

keeping it as-is in this PR to avoid breaking existing FT checkpoints.


> ### Design Consideration: `_async_wait` override
> **`torchtitan/experiments/ft/checkpoint.py:172-182`**
> 
> The FT subclass completely overrides `_async_wait` with different semantics — the base class distinguishes between `ASYNC_WITH_PINNED_MEM` and `ASYNC` (only the latter clears `save_future`), while the FT override unifies them into "clear unless ASYNC_WITH_PINNED_MEM." This works but the comment could note that FT's `_ft_save` always uses `AsyncMode.ASYNC` (line 200), so the DISABLED-with-save_future case only arises from `_ft_save`. This would make the override's rationale clearer.

Add comments: explains that _ft_save() always uses AsyncMode.ASYNC, which is why save_future can exist when
  self.async_mode is DISABLED.


> ### Test Coverage
> The PR removes `test_ft_async_save_calls_async_wait` and the `DummyFTManager` fixture from core tests, which is correct since FT logic no longer lives in core. However, **no corresponding FT tests are added** for the new `FTCheckpointManager`. The subclass has non-trivial logic:

Add test in `torchtitan/experiments/ft/tests/test_ft_checkpoint.py`

**Comment by mori360** [comment]
@claude review the code after addressing your comments

**Comment by mori360** [comment]
@claude review the code

File: .github/workflows/integration_test_8gpu_torchft.yaml [hunk 1]
```diff
@@ -81,6 +81,9 @@ jobs:
         sudo mkdir -p "$RUNNER_TEMP/artifacts-to-be-uploaded"
         sudo chown -R $(id -u):$(id -g) "$RUNNER_TEMP/artifacts-to-be-uploaded"
 
+        echo "ft_unit_tests"
+        pytest torchtitan/experiments/ft/tests/test_ft_checkpoint.py -v
```
Comment (fegin) [inline .github/workflows/integration_test_8gpu_torchft.yaml]: Do we need to add this unittest in 8GPU CIs? It only needs CPU.

Comment (tianyu-l) [inline .github/workflows/integration_test_8gpu_torchft.yaml]: this looks like a CPU test? don't add CPU test to this GPU integration test.

I don't mind we create a dedicated CPU test yaml workflow for it.

File: torchtitan/experiments/ft/trainer.py [hunk 2]
```diff
@@ -303,8 +303,11 @@ def __init__(self, config: Config):
         self.step = 0
         self.ntokens_seen = 0
 
-        # FT addition: pass ft_manager to CheckpointManager
-        self.checkpointer = config.checkpoint.build(
+        # FT addition: use FTCheckpointManager for fault tolerance support
+        from torchtitan.experiments.ft.checkpoint import FTCheckpointManager
```
Comment (tianyu-l) [inline torchtitan/experiments/ft/trainer.py]: don't need this import

File: torchtitan/experiments/ft/trainer.py [hunk 3]
```diff
@@ -303,8 +303,11 @@ def __init__(self, config: Config):
         self.step = 0
         self.ntokens_seen = 0
 
-        # FT addition: pass ft_manager to CheckpointManager
-        self.checkpointer = config.checkpoint.build(
```
Comment (tianyu-l) [inline torchtitan/experiments/ft/trainer.py:307]: no need to change code here -- the checkpointer is already a ft checkpoint manager config.

**Comment by mori360** [comment]
> Do you run real checkpointing integration tests? Run 50 steps and save and load and continue to run until 100 steps is finished?

Thanks for the comments, add test results in the description.

File: .github/workflows/integration_test_8gpu_torchft.yaml (hunk 1 above)
Comment (mori360) [reply .github/workflows/integration_test_8gpu_torchft.yaml]: Create `unit_test_cpu_torchft.yaml` to place the torchft cpu test.

File: .github/workflows/unit_test_cpu_torchft.yaml [hunk 4]
```diff
@@ -0,0 +1,41 @@
+name: TorchFT CPU Unit Test
+
+on:
+  push:
+    branches: [ main ]
+    paths:
+      - 'torchtitan/experiments/ft/**'
+      - 'torchtitan/components/checkpoint.py'
```
Comment (tianyu-l) [inline .github/workflows/unit_test_cpu_torchft.yaml]: let's not include this (core) in experiments development

File: .github/workflows/unit_test_cpu_torchft.yaml [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+      - '.github/workflows/unit_test_cpu_torchft.yaml'
+  pull_request:
+    paths:
+      - 'torchtitan/experiments/ft/**'
+      - 'torchtitan/components/checkpoint.py'
```
Comment (tianyu-l) [inline .github/workflows/unit_test_cpu_torchft.yaml]: same

**Review by tianyu-l — COMMENTED** [review]
> Test the main checkpoint by training 50 steps first, save, then load and train the next 50 steps.

In addition to this, we need to check that continue previous training gives the same results as save & load.
- run 50 steps, save and load, train the next 50
- vs. train 100 steps

**Comment by mori360** [comment]
> > [rank0]:[titan] 2026-04-04 22:39:24,887 - root - INFO - step: 50  loss:  2.80785  grad_norm:  0.2986  memory:  0.96GiB(1.01%)  tps: 159,509  tflops: 11.42  mfu: 1.15%
> 
> Your step 50 loss on "0->100" job is clearly different from "0->50" job. How did it achieve bitwise identity on the 60 to 100 steps???
> 
> In particular, the `training.steps` change will affect learning rate. I don't think you can hope for bitwise match in your current setup.

Yeah, 

> > [rank0]:[titan] 2026-04-04 22:39:24,887 - root - INFO - step: 50  loss:  2.80785  grad_norm:  0.2986  memory:  0.96GiB(1.01%)  tps: 159,509  tflops: 11.42  mfu: 1.15%
> 
> Your step 50 loss on "0->100" job is clearly different from "0->50" job. How did it achieve bitwise identity on the 60 to 100 steps???
> 
> In particular, the `training.steps` change will affect learning rate. I don't think you can hope for bitwise match in your current setup.

Yeah, the prev run had different loss among 11-50 step. Adding "--lr_scheduler.total_steps 100" to the first 50 steps running could get consistent loss.
